### PR TITLE
Add UTF8 DateTime(Offset) parser for new date and time format "I"

### PR DESCRIPTION
**Pull request by layomia**
This addresses parts of #34690. The  parser will be utilized by `Utf8JsonReader` to parse strings as `DateTime`.

The standard date and time format strings do not include a token "I". This token is now reserved to represent date and time strings that follow one of these variants of the ISO 8601 standard:

YYYY-MM-DD (eg 1997-07-16)
YYYY-MM-DDThh:mm (eg 1997-07-16T19:20)
YYYY-MM-DDThh:mm:ss (eg 1997-07-16T19:20:30)
YYYY-MM-DDThh:mm:ss.s (eg 1997-07-16T19:20:30.45)
YYYY-MM-DDThh:mmTZD (eg 1997-07-16T19:20+01:00)
YYYY-MM-DDThh:mm:ssTZD (eg 1997-07-16T19:20:30+01:00)
YYYY-MM-DDThh:mm:ss.sTZD (eg 1997-07-16T19:20:30.45+01:00)

where:
    YYYY = four-digit year
     MM   = two-digit month (01=January, etc.)
     DD   = two-digit day of month (01 through 31)
     hh   = two digits of hour (00 through 23) (am/pm NOT allowed)
     mm   = two digits of minute (00 through 59)
     ss   = two digits of second (00 through 59)
     s    = one or more digits representing a decimal fraction of a second
     TZD  = time zone designator (Z or +hh:mm or -hh:mm)

If specified, the decimal fraction of a second must contain >= 1 digit, but the parser will truncate digits after the 7th decimal place.

The parser validates input strings according to these formats and returns the corresponding `DateTime` values.

### CODE REVIEW

**Comment by layomia** [comment]
Tests for this change are in #35756.

**Comment by layomia** [comment]
@Tornhoof FYI

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+                            offsetChar = Utf8Constants.Plus;
+                            state = Utf8Constants.DateJParserState.OffsetHours;
+							sourceIndex++;
+                        }
+                        else if (source[sourceIndex] == Utf8Constants.Minus)
+                        {
+                            offsetChar = Utf8Constants.Minus;
+                            state = Utf8Constants.DateJParserState.OffsetHours;
+							sourceIndex++;
+						}
```
Comment (Tornhoof) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs]: I think there are a few whitespace issues in this file, you might want to reformat it.

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+					case Utf8Constants.DateJParserState.Minute:
+						if (!ParserHelpers.TryGetNextTwoDigits(source, sourceIndex, out sourceIndex, out minute))
+                        {
+                            state = Utf8Constants.DateJParserState.Invalid;
+                            break;
+                        }
+
+                        if (sourceLength - sourceIndex > 1)
+                        {
+                            if (source[sourceIndex] == Utf8Constants.Colon)
```
Comment (Tornhoof) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs]: A few lines above there is a switch for source[sourceIndex], why if/else now?

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+							{
+								continueParsing = false;
+								break;
+							}
+
+                            sourceIndex++;
+                        }
+						break;
+					case Utf8Constants.DateJParserState.Fraction:
+						while (sourceIndex < sourceLength && ParserHelpers.IsDigit(source[sourceIndex]))
```
Comment (Tornhoof) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs]: There is pretty much the same code in https://github.com/dotnet/coreclr/blob/5786f5b2f1b25ad0ad696f3cf248b6ca299aa356/src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.TimeSpan.cs#L79, maybe share it? The only difference is truncate vs abort? On the other hand if you plan to implement 'J' for timespan too, then you could reuse that code path.

**Comment by ahsonkhan** [comment]
> This addresses parts of #34690

https://github.com/dotnet/corefx/issues/34690

> Tests for this change are in #35756.

https://github.com/dotnet/corefx/pull/35756

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.cs [hunk 4]
```diff
@@ -89,6 +89,32 @@ public static bool TryParse(ReadOnlySpan<byte> source, out DateTime value, out i
                         return true;
                     }
 
+                case 'J':
+                    {
+                        if (!TryParseDateTimeOffsetJ(source, out DateTimeOffset dateTimeOffset, out bytesConsumed, out DateTimeKind kind))
```
Comment (ahsonkhan) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.cs]: Is this for Date or DateTimeOffset? The name is misleading.

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.cs [hunk 5]
```diff
@@ -89,6 +89,32 @@ public static bool TryParse(ReadOnlySpan<byte> source, out DateTime value, out i
                         return true;
                     }
 
+                case 'J':
```
Comment (ahsonkhan) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.cs]: Can you please update the xml docs for this API:
https://github.com/dotnet/coreclr/blob/fd3afcaf0dc0ace3f84d0412c128f4e64c419105/src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.cs#L22-L29

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+		// Flexible ISO 8601 format. One of
+		//
+		// ---------------------------------
+		// YYYY-MM-DD (eg 1997-07-16)
+		// YYYY-MM-DDThh:mm (eg 1997-07-16T19:20)
+		// YYYY-MM-DDThh:mm:ss (eg 1997-07-16T19:20:30)
+		// YYYY-MM-DDThh:mm:ss.s (eg 1997-07-16T19:20:30.45)
+		// YYYY-MM-DDThh:mmTZD (eg 1997-07-16T19:20+01:00)
+		// YYYY-MM-DDThh:mm:ssTZD (eg 1997-07-16T19:20:30+01:00)
+		// YYYY-MM-DDThh:mm:ss.sTZD (eg 1997-07-16T19:20:30.45+01:00)
```
Comment (ahsonkhan) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs]: nit: also include an example ending with `Z`.

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/ParserHelpers.cs [hunk 7]
```diff
@@ -70,5 +70,29 @@ public static bool TryParseThrowFormatException<T>(out T value, out int bytesCon
             value = default;
             return TryParseThrowFormatException(out bytesConsumed);
         }
+
+        public static bool TryGetNextTwoDigits(ReadOnlySpan<byte> source, int prevSourceIndex, out int sourceIndex, out int value)
```
Comment (ahsonkhan) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/ParserHelpers.cs]: Consider passing a sliced span instead. It is difficult to follow `prevSourceIndex`/`sourceIndex`
Alternatively, pass `sourceIndex` by ref (instead of having both `prevSourceIndex` and having an out).

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+					case Utf8Constants.DateJParserState.Second:
+						if (!ParserHelpers.TryGetNextTwoDigits(source, sourceIndex, out sourceIndex, out second))
+                        {
+                            state = Utf8Constants.DateJParserState.Invalid;
+                            break;
+                        }
+
+						if (sourceLength - sourceIndex > 1)
+                        {
+                            if (source[sourceIndex] == Utf8Constants.Period)
```
Comment (ahsonkhan) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs]: inconsistent braces - for nested scopes like this, add braces even for single line branches

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+					case Utf8Constants.DateJParserState.OffsetMinutes:
+						if (!ParserHelpers.TryGetNextTwoDigits(source, sourceIndex, out sourceIndex, out offsetMinutes))
+                        {
+                            state = Utf8Constants.DateJParserState.Invalid;
+                            break;
+                        }
+
+                        continueParsing = false;
+						break;
+					case Utf8Constants.DateJParserState.Invalid:
```
Comment (ahsonkhan) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs]: Do we really need the invalid state? Can we instead return false whenever an invalid state is detected?

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+					default:
```
Comment (ahsonkhan) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs]: What state would this catch? We may want to `Debug.Fail` for `default` since it likely indicates a bug in the code, no?

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+					bytesConsumed = 0;
+					kind = default;
+					return false;
+				}
+
+				kind = DateTimeKind.Unspecified;
+				return true;
+			}
+
+			if (offsetChar == 'Z')
```
Comment (ahsonkhan) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs]: Use `UtcOffsetChar`

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+
+			switch(state)
+			{
+				case Utf8Constants.DateJParserState.Day:
+				case Utf8Constants.DateJParserState.Minute:
+				case Utf8Constants.DateJParserState.Second:
+				case Utf8Constants.DateJParserState.Fraction:
+				case Utf8Constants.DateJParserState.OffsetMinutes:
+					break;
+				default:
```
Comment (ahsonkhan) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs]: What states is this `default` capturing?

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Constants.cs [hunk 13]
```diff
@@ -14,12 +14,14 @@ internal static partial class Utf8Constants
         public const byte Slash = (byte)'/';
         public const byte Space = (byte)' ';
         public const byte Hyphen = (byte)'-';
+		public const byte UtcOffsetChar = (byte)'Z';
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Constants.cs]: Nit: this includes tabs... please use spaces instead

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/ParserHelpers.cs [hunk 14]
```diff
@@ -70,5 +70,29 @@ public static bool TryParseThrowFormatException<T>(out T value, out int bytesCon
...
+        {
+            if (source.Length - prevSourceIndex < 2)
+            {
+                sourceIndex = prevSourceIndex;
+                value = default;
+                return false;
+            }
+
+            uint digit1 = source[prevSourceIndex++] - 48u;
+            uint digit2 = source[prevSourceIndex++] - 48u;
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/ParserHelpers.cs]: Nit: please use (uint)'0' instead; it's otherwise not clear what is the relevance of "48"

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs]: Yes, it's a bit hard to follow with the indentation being non-standard.  Thanks.

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs [hunk 15]
```diff
@@ -0,0 +1,298 @@
...
+							sourceIndex++;
+                        }
+                        else if (source[sourceIndex] == Utf8Constants.Minus)
+                        {
+                            offsetChar = Utf8Constants.Minus;
+                            state = Utf8Constants.DateJParserState.OffsetHours;
+							sourceIndex++;
+						}
+                        else
+                            continueParsing = false;
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs]: Nit: use braces

**Comment by stephentoub** [comment]
> The standard date and time format strings do not include a token "J". This token is now reserved to represent date and time strings that follow one of these variants of the ISO 8601 standard

If we're adding this to Utf8Parser, do we also need/want to:
- Add this format as well to Utf8Formatter?
- Add this format to DateTime{Offset}.{Try}Parse/Format/ToString?

**Comment by ahsonkhan** [comment]
> Add this format as well to Utf8Formatter?

Yes, particularly to support https://github.com/dotnet/corefx/issues/34576, which is very similar to the existing 'O' format.

> Add this format to DateTime{Offset}.{Try}Parse/Format/ToString?

Imo, for `{Try}ParseExact`, we should consider it. However, it isn't as necessary since `DateTime.Parse` is already forgiving and accepts a variety of inputs by default.

@JamesNK, @jkotas what are your thoughts around adding 'J' to DateTime.Format/ToString matching the semantics of the `Utf8Formatter` (i.e. remove trailing zeros on the fractional component)?

**Comment by jkotas** [comment]
>  what are your thoughts around adding 'J' to DateTime.Format/ToString

I do not have strong opinion myself. This is in essence a new public API. It should have API proposal issues that should be reviewed just like any other new public API.

**Comment by ahsonkhan** [comment]
> This is in essence a new public API. It should have API proposal issues that should be reviewed just like any other new public API.

@layomia, can you file an issue for adding 'J' support to DateTime.ParseExact/Format/etc. (in corefx)?

Take a look at the doc: https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md (with sample issue style - https://github.com/dotnet/corefx/issues/271).

**Comment by Tornhoof** [comment]
> @layomia, can you file an issue for adding 'J' support to DateTime.ParseExact/Format/etc. (in corefx)?
I have a previous one open for that https://github.com/dotnet/corefx/issues/30084

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs (hunk 3 above)
Comment (layomia) [reply src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs]: The `TryParseTimeSpanFraction` method does not return the parsed value when it aborts. Will leave as is for now.

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs [hunk 16]
```diff
@@ -0,0 +1,388 @@
...
+			if (sourceLength - sourceIndex < 5)
+			{
+				value = default;
+				bytesConsumed = 0;
+				kind = default;
+				return false;
+			}
+
+			{
+				uint digit1 = source[sourceIndex++] - (uint)'0';
```
[the remark this answers is not in the document]
Comment (layomia) [reply src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs]: Doesn't regress perf.

**Comment by layomia** [comment]
Benchmark on my machine for `DateTime.TryParseExact` (baseline), Utf8Parser.Date.J, and Utf8Parser.Date.O for input string "1997-07-16T19:20:33.3333333+01:00":

``` ini

BenchmarkDotNet=v0.11.4, OS=Windows 10.0.17763.316 (1809/October2018Update/Redstone5)
Intel Core i7-7700 CPU 3.60GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-preview-010184
  [Host]     : .NET Core 3.0.0-preview-27324-5 (CoreCLR 4.6.27322.0, CoreFX 4.7.19.7311), 64bit RyuJIT
  Job-HQHPYX : .NET Core ca33148e-5831-4534-afc7-f8d27b6d302b (CoreCLR 4.6.27505.0, CoreFX 4.7.19.15601), 64bit RyuJIT

Runtime=Core  Toolchain=CoreRun  

```
|                      Method |           TestString |      Mean |     Error |    StdDev | Ratio | RatioSD | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|---------------------------- |--------------------- |----------:|----------:|----------:|------:|--------:|------------:|------------:|------------:|--------------------:|
| **DateTimeTryParseExact** | **1997-(...)01:00 [33]** | **760.49 ns** | **5.4621 ns** | **5.1093 ns** | **762.31 ns** |  **1.00** |      **0.0286** |           **-** |           **-** |               **120 B** |
|   Utf8DateTimeParserJ | 1997-(...)01:00 [33] | 272.66 ns | 2.0866 ns | 1.9518 ns | 273.66 ns |  0.36 |           - |           - |           - |                   - |
|   Utf8DateTimeParserO | 1997-(...)01:00 [33] | 263.96 ns | 5.2010 ns | 4.8650 ns | 263.04 ns |  0.35 |           - |           - |           - |                   - |
| **DateTimeOffsetTryParseExact** | **1997-(...)01:00 [33]** | **107.05 ns** | **0.9533 ns** | **0.8917 ns** |  **1.00** |    **0.00** |           **-** |           **-** |           **-** |                   **-** |
|   Utf8DateTimeOffsetParserJ | 1997-(...)01:00 [33] |  58.97 ns | 0.3270 ns | 0.3059 ns |  0.55 |    0.00 |           - |           - |           - |                   - |
|   Utf8DateTimeOffsetParserO | 1997-(...)01:00 [33] |  47.58 ns | 0.4886 ns | 0.4570 ns |  0.44 |    0.01 |           

```C#
using System;
using System.Buffers.Text;
using System.Globalization;
using System.Text;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Diagnosers;
using BenchmarkDotNet.Running;

namespace MyBenchMark
{
	[MemoryDiagnoser]
	public class Utf8DateTimeParserBenchmark
	{
		[Params("1997-07-16T19:20:33.3333333+01:00")]
		public string TestString;
		byte[] Utf8ByteArray;

		[GlobalSetup]
		public void Setup()
		{
			Utf8ByteArray = Encoding.UTF8.GetBytes(TestString);
		}

		[Benchmark(Baseline = true)]
		public DateTime DateTimeTryParseExact()
		{
			ReadOnlySpan<byte> utf8ByteSpan = Utf8ByteArray;
			DateTime.TryParseExact(TestString, "O", CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind, out DateTime value);
			return value;
		}

		[Benchmark]
		public DateTime Utf8DateTimeParserJ()
		{
			ReadOnlySpan<byte> utf8ByteSpan = Utf8ByteArray;
			Utf8Parser.TryParse(utf8ByteSpan, out DateTime value, out int bytesConsumed, 'J');
			return value;
		}

		[Benchmark]
		public DateTime Utf8DateTimeParserO()
		{
			ReadOnlySpan<byte> utf8ByteSpan = Utf8ByteArray;
			Utf8Parser.TryParse(utf8ByteSpan, out DateTime value, out int bytesConsumed, 'O');
			return value;
		}
	}

	[MemoryDiagnoser]
	public class Utf8DateTimeOffsetParserBenchmark
	{
		[Params("1997-07-16T19:20:33.3333333+01:00")]
		public string TestString;
		byte[] Utf8ByteArray;

		[GlobalSetup]
		public void Setup()
		{
			Utf8ByteArray = Encoding.UTF8.GetBytes(TestString);
		}

		[Benchmark(Baseline = true)]
		public DateTimeOffset DateTimeOffsetTryParseExact()
		{
			ReadOnlySpan<byte> utf8ByteSpan = Utf8ByteArray;
			DateTimeOffset.TryParseExact(TestString, "O", CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind, out DateTimeOffset value);
			return value;
		}

		[Benchmark]
		public DateTimeOffset Utf8DateTimeOffsetParserJ()
		{
			ReadOnlySpan<byte> utf8ByteSpan = Utf8ByteArray;
			Utf8Parser.TryParse(utf8ByteSpan, out DateTimeOffset value, out int bytesConsumed, 'J');
			return value;
		}

		[Benchmark]
		public DateTimeOffset Utf8DateTimeOffsetParserO()
		{
			ReadOnlySpan<byte> utf8ByteSpan = Utf8ByteArray;
			Utf8Parser.TryParse(utf8ByteSpan, out DateTimeOffset value, out int bytesConsumed, 'O');
			return value;
		}
	}

        public class Program
        {
                static void Main(string[] args) => BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args);
        }
}
```

**Comment by jkotas** [comment]
> I have a previous one open for that dotnet/corefx#30084

That issue is marked for "Future" and it is not in the right format for the API review...

**Comment by JamesNK** [comment]
Is the "J" for JSON? If the format is for ISO8601 then I think "I" would be better. That follows what the existing strings do - https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-date-and-time-format-strings

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.I.cs [hunk 17]
```diff
@@ -0,0 +1,305 @@
...
+            if (curByte == Utf8Constants.UtcOffsetToken)
+            {
+                bytesConsumed++;
+                offsetToken = Utf8Constants.UtcOffsetToken;
+                goto FinishedParsing;
+            }
+            else if (curByte == Utf8Constants.Plus || curByte == Utf8Constants.Minus)
+            {
+                offsetToken = curByte;
+                sourceIndex += 1;
```
Comment (ahsonkhan) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.I.cs]: nit: sourceIndex++

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs [hunk 18]
```diff
@@ -196,16 +198,17 @@ private static bool TryParseDateTimeOffsetJ(ReadOnlySpan<byte> source, out DateT
...
+					sourceIndex++;
 				}
 			}
 
 			if (fraction != 0)
 			{
-				int fractionTimesTen;
-				while ((fractionTimesTen = fraction * 10) <= Utf8Constants.MaxDateTimeFraction)
+				// Note this is 1 order of magnitude less than Utf8Constants.MaxDateTimeFraction
+				while (fraction <= 999_999)
```
Comment (ahsonkhan) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.J.cs]: nit: use named constant for this. Maybe `MaxDateTimeFractionDiv10` or something.

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.I.cs [hunk 19]
```diff
@@ -0,0 +1,305 @@
...
+
+                    fraction = prevFractionTimesTen + (int)(curByte - (uint)'0');
+                    sourceIndex++;
+                }
+            }
+
+            if (fraction != 0)
+            {
+                // Note this is 1 order of magnitude less than Utf8Constants.MaxDateTimeFraction
+                while (fraction <= 999_999)
```
Comment (ahsonkhan) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.I.cs]: nit: use named constant for this. Maybe `MaxDateTimeFractionDiv10` or something.

**Comment by layomia** [comment]
> Looks good, pending the change around
> 
> ```
> <time>Z
> <time>±hh:mm
> <time>±hhmm
> <time>±hh
> ```
> [dotnet/corefx#35756 (comment)](https://github.com/dotnet/corefx/pull/35756#discussion_r263080986)

Logged an issue (https://github.com/dotnet/corefx/issues/35845), will be addressed in subsequent PR.

File: src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Constants.cs [hunk 20]
```diff
@@ -14,6 +14,9 @@ internal static partial class Utf8Constants
         public const byte Slash = (byte)'/';
         public const byte Space = (byte)' ';
         public const byte Hyphen = (byte)'-';
+        public const byte Hyphen = (byte)'-';
```
Comment (ahsonkhan) [inline src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Constants.cs]:
```text
error CS0102: The type 'Utf8Constants' already contains a definition for 'Hyphen' 
```

**Comment by layomia** [comment]
Given feedback, introducing a new format specifier, especially one that enables ISO 8601, requires further investigation and design. https://github.com/dotnet/corefx/issues/35976 tracks this discussion.
